textattack/transformations/word_swap_masked_lm.py

```diff
@@ -0,0 +1,209 @@
+        self.method = method
+        self.masked_lm_name = masked_language_model
+        self.max_length = max_length
+        self.max_candidates = max_candidates
+
+        self._lm_tokenizer = AutoTokenizer.from_pretrained(
+            masked_language_model, use_fast=True
+        )
+        self._language_model = AutoModelWithLMHead.from_pretrained(
+            masked_language_model
+        )
+        self._language_model.to(utils.device)
+        self._segment_tensor = (
+            torch.zeros(self.max_length, dtype=torch.long).unsqueeze(0).to(utils.device)
+        )
+        self._language_model.eval()
+
+    def _bae_replacement_words(self, current_text, index):
+        # TODO: is it necessary to create a new AttackedText to recover case and punctuation?
+        masked_attacked_text = current_text.replace_word_at_index(
+            index, self._lm_tokenizer.mask_token
+        )
+        ids = self._lm_tokenizer.encode(
+            masked_attacked_text.text,
+            max_length=self.max_length,
+            pad_to_max_length=True,
+        )
+
+        try:
+            masked_index = ids.index(self._lm_tokenizer.mask_token_id)
+        except ValueError:
+            return []
+
+        ids_tensor = torch.tensor([ids]).to(utils.device)
+        with torch.no_grad():
+            preds = self._language_model(
+                ids_tensor, token_type_ids=self._segment_tensor
+            )[0]
+
+        mask_token_probs = preds[0, masked_index]
+        topk = torch.topk(mask_token_probs, self.max_candidates)
+        top_logits = topk[0].tolist()
+        top_ids = topk[1].tolist()
+
+        replacement_words = []
+        for id in top_ids:
+            token = self._lm_tokenizer.convert_ids_to_tokens(id)
+            if check_if_word(token):
+                replacement_words.append(token)
+
+        return replacement_words
+
+    def _bert_attack_replacement_words(self, current_text, index, extra_args):
+        top_pred_ids = extra_args["top_pred_ids"]
+        masked_lm_logits = extra_args["masked_lm_logits"]
+        current_ids = extra_args["current_ids"]
+        token2char_offset = extra_args["token2char_offset"]
+
+        target_word = current_text.words[index]
+        # Start and end position of word we want to replace in the whole text
+        word_start, word_end = current_text.words2char_offset[target_word]
+
+        # We need to find which BPE tokens belong to the word we want to replace
+        # List of indices of tokens that are part of the target word
+        target_ids_pos = []
+
+        for i in range(len(current_ids)):
+            token_start, token_end = token2char_offset[i]
+            token = self._lm_tokenizer.convert_ids_to_tokens(current_ids[i]).replace(
+                "##", ""
+            )
+            if (
+                token_start >= word_start and token_end - 1 <= word_end
+            ) and token in target_word:
+                target_ids_pos.append(i)
```

Comment:
I see the problem here but I'm not convinced this is the best solution. Maybe you could encode the first part of the string and then take all the tokens after that that form a single word? (As in, take ['word'] or ['wo##' '##rd'], depending on which you find.) Maybe I'm missing why this is a technical issue.

If you can't think of a better way then maybe just put a @TODO to clean this up later :)

Reply:
I do have another way of doing this. Mask out the word we want to replace like in BAE function, then after tokenizing, replace Mask token id with the tokens of word we want to replace. Then, we know exactly which tokens belong to the word we want to replace. Idk if this is the best way, however, since we end up tokenizing the same text over and over again. But then again, it might not matter much since we already go through the list of ids once already.

Reply:
thats a very good idea :)